File: packages/duoduo/src/smart-layer/ast.ts
import { SmartLayerClient } from "./client"

// ─── Types ────────────────────────────────────────────────────────────────

export interface ValidateSyntaxRequest {
  code: string
  language: string
}

/** Wire format of `duo_types::FunctionDef` (`common.rs:701-713`, camelCase). */
export interface FunctionDef {
  name: string
  returnType?: string
  startLine: number
  endLine: number
  parameters?: string[]
  documentation?: string
}

export interface ValidateSyntaxResponse {
  valid: boolean
  errors: string[]
  /** `FunctionDef[]` objects, NOT strings — the old `string[]` declaration
   * never matched the payload. */
  functions: FunctionDef[]
  language: string
}

/**
 * Wire format of `StructuralDiffRequest` (`routes/ast.rs:30-37`, NO
 * `rename_all` ⇒ snake_case keys). `agentId` is a required `String` on the
 * Rust side — the old camelCase keys made every call 400.
 */
export interface StructuralDiffRequest {
  file_path: string
  old_code: string
  new_code: string
  language: string
  agent_id: string
}

/** Wire format of `FileChangeEntry` (`duo-types/blackboard.rs:63-69`, no
 * rename ⇒ snake_case). */
export interface StructuredChange {
  symbol_name: string
  change_kind: string
  old_signature: string
  new_signature: string
}

/** Wire format of `StructuredChangeList` (`blackboard.rs:72-77`). */
export interface StructuralDiffResponse {
  file: string
  agent_id: string
  changes: StructuredChange[]
}

export interface ComputeHashRequest {
  code: string
  language: string
}
export interface ComputeHashResponse {
  hash: string
}

// ─── Client ───────────────────────────────────────────────────────────────

/**
 * AstClient provides typed access to the AST analysis API.
 *
 * Supports:
 *   - validateSyntax: check code syntax and extract functions
 *   - structuralDiff: compute structured changes between two code versions
 *   - computeHash: compute an AST-based hash for change detection
 */
export class AstClient {
  constructor(private client: SmartLayerClient) {}

  /** Validate code syntax and extract function names. */
  async validateSyntax(req: ValidateSyntaxRequest): Promise<ValidateSyntaxResponse> {
    return this.client.post<ValidateSyntaxResponse>("/ast/validate-syntax", req)
  }

  /** Compute structured changes between two code versions. */
  async structuralDiff(req: StructuralDiffRequest): Promise<StructuralDiffResponse> {
    return this.client.post<StructuralDiffResponse>("/ast/structural-diff", req)
  }

  /** Compute an AST-based hash for change detection. */
  async computeHash(req: ComputeHashRequest): Promise<ComputeHashResponse> {
    return this.client.post<ComputeHashResponse>("/ast/compute-hash", req)
  }
}

File: crates/duo-smart-layer/src/routes/ast.rs
//! AST routes — syntax validation, structural diff, and AST hash computation.

use axum::{Json, Router};

use crate::error::{Result, bad_request};

pub fn router() -> Router<crate::server::AppState> {
    Router::new()
        .route("/ast/validate-syntax", axum::routing::post(validate_syntax))
        .route("/ast/structural-diff", axum::routing::post(structural_diff))
        .route("/ast/compute-hash", axum::routing::post(compute_hash))
}

// ─── Request / Response types ────────────────────────────────────────────

#[derive(serde::Deserialize)]
struct ValidateSyntaxRequest {
    code: String,
    language: String,
}

#[derive(serde::Serialize)]
struct ValidateSyntaxResponse {
    valid: bool,
    errors: Vec<String>,
    functions: Vec<duo_types::FunctionDef>,
    language: String,
}

#[derive(serde::Deserialize)]
struct StructuralDiffRequest {
    file_path: String,
    old_code: String,
    new_code: String,
    language: String,
    agent_id: String,
}

#[derive(serde::Deserialize)]
struct ComputeHashRequest {
    code: String,
    language: String,
}

#[derive(serde::Serialize)]
struct ComputeHashResponse {
    hash: String,
}

// ─── Fallback helpers ────────────────────────────────────────────────────

/// Fallback content hash when tree-sitter is unavailable for a language.
fn fallback_content_hash(code: &str) -> String {
    use std::hash::{Hash, Hasher};
    let stripped: String = code.chars().filter(|c| !c.is_whitespace()).collect();
    let mut hasher = std::collections::hash_map::DefaultHasher::new();
    stripped.hash(&mut hasher);
    format!("{:016x}", hasher.finish())
}

/// Empty change list for when structural diff is unavailable.
fn empty_change_list(file_path: &str, agent_id: &str) -> duo_types::StructuredChangeList {
    duo_types::StructuredChangeList {
        file: file_path.to_string(),
        agent_id: agent_id.to_string(),
        changes: vec![],
    }
}

// ─── Handlers ─────────────────────────────────────────────────────────────

/// Validate code syntax without writing to disk.
async fn validate_syntax(
    Json(req): Json<ValidateSyntaxRequest>,
) -> Result<Json<ValidateSyntaxResponse>> {
    if req.code.is_empty() {
        return Err(bad_request(anyhow::anyhow!("code must not be empty")));
    }

    let result =
        tokio::task::spawn_blocking(move || ast_engine::analyze(&req.code, &req.language)).await?;

    Ok(Json(ValidateSyntaxResponse {
        valid: true,
        errors: result.errors.unwrap_or_default(),
        functions: result.functions,
        language: result.language,
    }))
}

/// Generate a structural diff between two versions of a file.
/// Falls back to an empty change list when tree-sitter is unavailable.
async fn structural_diff(
    Json(req): Json<StructuralDiffRequest>,
) -> Result<Json<duo_types::StructuredChangeList>> {
    let result = tokio::task::spawn_blocking(move || {
        if ast_engine::is_language_enabled(&req.language) {
            ast_engine::generate_structural_diff(
                &req.file_path,
                &req.old_code,
                &req.new_code,
                &req.language,
                &req.agent_id,
            )
            .unwrap_or_else(|_| empty_change_list(&req.file_path, &req.agent_id))
        } else {
            empty_change_list(&req.file_path, &req.agent_id)
        }
    })
    .await?;

    Ok(Json(result))
}

/// Compute a deterministic AST hash for source code.
/// Falls back to a content-based hash when tree-sitter is unavailable.
async fn compute_hash(Json(req): Json<ComputeHashRequest>) -> Result<Json<ComputeHashResponse>> {
    if req.code.is_empty() {
        return Err(bad_request(anyhow::anyhow!("code must not be empty")));
    }

    let hash = tokio::task::spawn_blocking(move || {
        if ast_engine::is_language_enabled(&req.language) {
            ast_engine::compute_ast_hash(&req.code, &req.language)
                .unwrap_or_else(|_| fallback_content_hash(&req.code))
        } else {
            fallback_content_hash(&req.code)
        }
    })
    .await?;

    Ok(Json(ComputeHashResponse { hash }))
}
